### PR TITLE
Convert the SSLv3/TLSv1.0/1.1/1.2 crypto code to use the new write record layer

### PR DESCRIPTION
**Pull request by mattcaswell**
This is built on top of #19198 and extends it to move the TLSv1.0/1.1/1.2 crypto code into the new write record layer.

Only the last commit is relevant. All the rest are inherited from #19198. For that reason this will remain in draft until that one has been merged.

### CODE REVIEW

**Comment by slontis** [comment]
Are there are performance comparisons being done as part of this work?

**Comment by mattcaswell** [comment]
> Are there are performance comparisons being done as part of this work?

No. There have been separate ongoing discussions about having some performance testing infrastructure which would allow us to spot and track performance changes over time - but nothing concrete as yet. Testing performance is probably premature at this stage since the refactor is incomplete.

**Comment by mattcaswell** [comment]
I've pushed an additional commit to address the CI failure. So now the last 2 commits in this PR are relevant and all earlier ones are inherited from #19198.

**Comment by mattcaswell** [comment]
I've extended this to also convert the SSLv3 code

**Comment by mattcaswell** [comment]
Now that #19198 has been merged I have rebased this PR and taken it out of draft.

There is a CI failure that appears unrelated in bio_dgram_test (ping @hlandau).

Please review!

**Comment by hlandau** [comment]
Needs rebase

**Comment by mattcaswell** [comment]
Rebased

File: ssl/record/methods/tls1_meth.c [hunk 1]
```diff
@@ -118,6 +119,29 @@ static int tls1_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
...
 
+    /* Calculate the explict IV length */
+    if (RLAYER_USE_EXPLICIT_IV(rl)) {
+        int mode = EVP_CIPHER_CTX_get_mode(ciph_ctx);
+        int eivlen = 0;
+
+        if (mode == EVP_CIPH_CBC_MODE) {
+            eivlen = EVP_CIPHER_CTX_get_iv_length(ciph_ctx);
+            if (eivlen < 0) {
+                RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, SSL_R_LIBRARY_BUG);
```
Comment (hlandau) [inline ssl/record/methods/tls1_meth.c:130]: nit: indentation

File: ssl/record/methods/tls1_meth.c [hunk 2]
```diff
@@ -118,6 +119,29 @@ static int tls1_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
...
+               return OSSL_RECORD_RETURN_FATAL;
+
```
Comment (hlandau) [inline ssl/record/methods/tls1_meth.c:132]: nit: extra blank line

File: ssl/record/methods/tls_common.c [hunk 3]
```diff
@@ -981,6 +982,24 @@ int tls_default_validate_record_header(OSSL_RECORD_LAYER *rl, SSL3_RECORD *rec)
...
+{
+#ifndef OPENSSL_NO_COMP
+    int i;
+
+    i = COMP_compress_block(rl->compctx, wr->data,
+                            (int)(wr->length + SSL3_RT_MAX_COMPRESSED_OVERHEAD),
+                            wr->input, (int)wr->length);
+    if (i < 0)
+        return 0;
+    else
```
Comment (hlandau) [inline ssl/record/methods/tls_common.c:995]: style: else is not needed here

**Review by hlandau — CHANGES_REQUESTED** [review]
Minor nits, other than that looks good.

File: ssl/record/methods/tls1_meth.c (hunk 2 above)
Comment (levitte) [reply ssl/record/methods/tls1_meth.c:132]: Also, the `return` line is misaligned by one space

File: ssl/record/methods/tls_multib.c [hunk 4]
```diff
@@ -113,7 +108,7 @@ static int tls_write_records_multiblock_int(OSSL_RECORD_LAYER *rl,
      * multiblock write in the call to tls_setup_write_buffer() - the different
      * buffer sizes will be spotted and the buffer reallocated.
      */
-    packlen = EVP_CIPHER_CTX_ctrl(s->enc_write_ctx,
+    packlen = EVP_CIPHER_CTX_ctrl(rl->enc_ctx,
                                     EVP_CTRL_TLS1_1_MULTIBLOCK_MAX_BUFSIZE,
                                     (int)templates[0].buflen, NULL);
```
Comment (levitte) [inline ssl/record/methods/tls_multib.c:113]: The arguments are misaligned.  Mind fixing it?

File: ssl/record/methods/tls_multib.c [hunk 5]
```diff
@@ -134,7 +129,7 @@ static int tls_write_records_multiblock_int(OSSL_RECORD_LAYER *rl,
     mb_param.inp = aad;
     mb_param.len = totlen;
 
-    packleni = EVP_CIPHER_CTX_ctrl(s->enc_write_ctx,
+    packleni = EVP_CIPHER_CTX_ctrl(rl->enc_ctx,
                                     EVP_CTRL_TLS1_1_MULTIBLOCK_AAD,
                                     sizeof(mb_param), &mb_param);
```
Comment (levitte) [inline ssl/record/methods/tls_multib.c:134]: The arguments are misaligned. Mind fixing it?

**Comment by mattcaswell** [comment]
All feedback addressed. @hlandau, @levitte - please take another look.

File: ssl/record/methods/tls1_meth.c [hunk 6]
```diff
@@ -118,6 +119,28 @@ static int tls1_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
...
+        int eivlen = 0;
+
+        if (mode == EVP_CIPH_CBC_MODE) {
+            eivlen = EVP_CIPHER_CTX_get_iv_length(ciph_ctx);
+            if (eivlen < 0) {
+                RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, SSL_R_LIBRARY_BUG);
+                return OSSL_RECORD_RETURN_FATAL;
+            }
+            if (eivlen <= 1)
+                eivlen = 0;
```
Comment (t8m) [inline ssl/record/methods/tls1_meth.c:134]: Why is this needed? Does it even make sense to have such short iv in EVP_CIPH_CBC_MODE?

File: ssl/record/methods/tls_common.c [hunk 7]
```diff
@@ -981,6 +982,23 @@ int tls_default_validate_record_header(OSSL_RECORD_LAYER *rl, SSL3_RECORD *rec)
...
+    i = COMP_compress_block(rl->compctx, wr->data,
+                            (int)(wr->length + SSL3_RT_MAX_COMPRESSED_OVERHEAD),
+                            wr->input, (int)wr->length);
+    if (i < 0)
+        return 0;
+
+    wr->length = i;
+    wr->input = wr->data;
+#endif
+    return 1;
```
Comment (t8m) [inline ssl/record/methods/tls_common.c:999]: Is this right return value? Can this be ever called with OPENSSL_NO_COMP?

**Review by t8m — COMMENTED** [review]
Just some questions to answer. Otherwise LGTM.

File: ssl/record/methods/tls1_meth.c (hunk 6 above)
Comment (t8m) [reply ssl/record/methods/tls1_meth.c:134]: OK then.

File: ssl/record/methods/tls_common.c (hunk 7 above)
Comment (mattcaswell) [reply ssl/record/methods/tls_common.c:999]: In practice it shouldn't make a difference since it can't be called in the case of `OPENSSL_NO_COMP`. However if something went wrong and it ended up being called anyway then it would be better if we default to return of 0 - in the same way that we do for `tls_do_uncompress`. I've fixed the code to do it that way.

**Comment by mattcaswell** [comment]
Fixup pushed addressing the feedback.

**Review by t8m — APPROVED** [review]

**Review by hlandau — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks. I fixed a trivial conflict during merge caused by some adjustment to error codes in code that was being deleted by this PR.

**Comment by wan7teen-ping** [comment]
Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Sab, 17 Sep 2022, 12:26 AM Matt Caswell ***@***.***>
menulis:
